### PR TITLE
add support for compile with dependencies for the maven plugin

### PR DESCRIPTION
**Pull request by guyhanael**

### CODE REVIEW

**Comment by guyhanael** [comment]
@Bonczidai is it possible to include this PR in the next minor release aka 0.9.72 ? I want to use it in the test content.

File: cloudslang-content-maven-compiler/src/main/java/io/cloudslang/maven/compiler/CloudSlangMavenCompiler.java [hunk 1]
```diff
@@ -3,36 +3,38 @@
...
 import org.springframework.context.ApplicationContext;
 import org.springframework.context.annotation.AnnotationConfigApplicationContext;
 
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.*;
```
Comment (Bonczidai) [inline cloudslang-content-maven-compiler/src/main/java/io/cloudslang/maven/compiler/CloudSlangMavenCompiler.java]: don't use `*` imports (there also an IntelliJ setting to change the default behavior)

File: cloudslang-content-maven-compiler/src/main/java/io/cloudslang/maven/compiler/CloudSlangMavenCompiler.java [hunk 2]
```diff
@@ -3,36 +3,38 @@
...
+import java.util.jar.JarFile;
+
 /**
  * Created by hanael on 10/07/2016.
  */
 
-@Component(role = org.codehaus.plexus.compiler.Compiler.class, hint = "cloudslang")
-public class CloudSlangMavenCompiler extends AbstractCompiler {
+@Component(role = org.codehaus.plexus.compiler.Compiler.class , hint = "cloudslang" )
+public class CloudSlangMavenCompiler extends AbstractCompiler{
```
Comment (Bonczidai) [inline cloudslang-content-maven-compiler/src/main/java/io/cloudslang/maven/compiler/CloudSlangMavenCompiler.java]: space between class name and `}` you can try the IntellJ `Reformat code` option on your classes

File: cloudslang-content-maven-compiler/src/main/java/io/cloudslang/maven/compiler/CloudSlangMavenCompiler.java [hunk 3]
```diff
@@ -48,86 +50,180 @@ public boolean canUpdateTarget(CompilerConfiguration configuration) throws Compi
...
 
         return compilerResult;
     }
 
-    private List<CompilerMessage> compileFile(String sourceFile) {
-        ExecutableModellingResult executableModellingResult;
-        List<CompilerMessage> compilerMessage = new ArrayList<>();
+    private void init(CompilerConfiguration config) {
+        //This parameter is passed in the compiler plugin whether to compile the flow with its dependencies
+        compileWithDependencies = !config.getCustomCompilerArgumentsAsMap().containsKey("ignore-dependencies");
```
Comment (Bonczidai) [inline cloudslang-content-maven-compiler/src/main/java/io/cloudslang/maven/compiler/CloudSlangMavenCompiler.java]: I would extract  a const

File: cloudslang-content-maven-compiler/src/main/java/io/cloudslang/maven/compiler/CloudSlangMavenCompiler.java [hunk 4]
```diff
@@ -48,86 +50,180 @@ public boolean canUpdateTarget(CompilerConfiguration configuration) throws Compi
...
+        //This parameter is used to control the error level. if not set only warnings will be shown
+        errorLevel = config.getCustomCompilerArgumentsAsMap().containsKey("ignore-errors") ? CompilerMessage.Kind.WARNING : CompilerMessage.Kind.ERROR;
```
Comment (Bonczidai) [inline cloudslang-content-maven-compiler/src/main/java/io/cloudslang/maven/compiler/CloudSlangMavenCompiler.java]: I would extract  a const

Comment (Bonczidai) [inline cloudslang-content-maven-compiler/src/main/java/io/cloudslang/maven/compiler/CloudSlangMavenCompiler.java]: consider breaking long lines (that does not fit review window - it's a good reference point)

File: cloudslang-content-maven-compiler/src/main/java/io/cloudslang/maven/compiler/CloudSlangMavenCompiler.java [hunk 5]
```diff
@@ -48,86 +50,180 @@ public boolean canUpdateTarget(CompilerConfiguration configuration) throws Compi
...
+            return compilerMessages ;
+        }
+
+        for (String dependency : dependencies) {
+            try {
+                SlangSource slangSource = SlangSource.fromFile(new File(dependency));
+                executableModellingResult = slangCompiler.preCompileSource(slangSource);
+                dependenciesExecutables.add(executableModellingResult.getExecutable());
+            } catch (Exception e){
+                //do nothing
```
Comment (Bonczidai) [inline cloudslang-content-maven-compiler/src/main/java/io/cloudslang/maven/compiler/CloudSlangMavenCompiler.java]: don;t we want to log at least here?

File: cloudslang-content-maven-compiler/src/main/java/io/cloudslang/maven/compiler/CloudSlangMavenCompiler.java [hunk 6]
```diff
@@ -48,86 +50,180 @@ public boolean canUpdateTarget(CompilerConfiguration configuration) throws Compi
...
+            }
+        }
+
+        for (Map.Entry<String, byte[]> dependencyEntry : dependenciesSourceFiles.entrySet()) {
+            try {
+                SlangSource slangSource = SlangSource.fromBytes(dependencyEntry.getValue(),dependencyEntry.getKey());
+                executableModellingResult = slangCompiler.preCompileSource(slangSource);
+                dependenciesExecutables.add(executableModellingResult.getExecutable());
+            } catch (Exception e){
+                //do nothing
```
Comment (Bonczidai) [inline cloudslang-content-maven-compiler/src/main/java/io/cloudslang/maven/compiler/CloudSlangMavenCompiler.java]: don't we want to log at least here?

**Comment by Bonczidai** [comment]
👍 

also here are a couple of tips:

- missing signed-off-by in PR description and is commit messages
- in the future consider creating a new feature branch in your fork (I am not much fan of master branch)

**Comment by Bonczidai** [comment]
also consider updating from master (origin)

File: cloudslang-content-maven-compiler/README.md [hunk 7]
```diff
@@ -18,6 +21,10 @@ The example below shows how to use the CloudSlang compiler:
         <version>3.5.1</version>
         <configuration>
           <compilerId>cloudslang</compilerId>
+           <compilerArgs>
+             <!--<arg>ignore-dependencies</arg> if you do not want to compile with dependnecies-->
+             <!--<arg>ignore-errors</arg> if you want to get only warnings and not fail the build -->
+           </compilerArgs>
         </configuration>
         <dependencies>
           <dependency>
```
Comment (Bonczidai) [inline cloudslang-content-maven-compiler/README.md]: ending tags for: 

```
    </plugin>
    </plugins>
    </build>
```

**Comment by Bonczidai** [comment]
maybe we cna consider an possible enhancement when displaying:

```
Compiling 449 source files
```

we could could display additional message for each compiled artifact real time;
also seems to me that the compilation takes relatively much time  ~ 20 / 30 secs - is it supposed to be like this?

**Comment by guyhanael** [comment]
@Bonczidai  - on what project did you try and use the plugin that it took a lot of time. projects with a lot of dependencies, I extract the jar files which take time.

**Comment by Bonczidai** [comment]
I've tried it out on our content repository

alright, merging 👍

**Pull request merged**